### PR TITLE
Add deploy to cloudflare worker documentation

### PR DESCRIPTION
**Pull request by matthewmueller**
## Describe this PR

This PR adds a guide for deploying to Cloudflare Workers with Prisma Client connecting through the Data Proxy to a MongoDB instance hosted on Atlas.

[Preview URL](https://deploy-preview-2404--prisma2-docs.netlify.app/docs/guides/deployment/deployment-guides/deploying-to-cloudflare-workers)

### CODE REVIEW

File: content/300-guides/200-deployment/110-deployment-guides/450-deploying-to-cloudflare-workers.mdx [hunk 1]
```diff
@@ -0,0 +1,322 @@
+---
+title: 'Deploying to Cloudflare Workers'
+metaTitle: 'Deploying to Cloudflare Workers'
+metaDescription: 'Learn how to deploy a TypeScript application to Cloudflare Workers that connects to MongoDB Atlas.'
+---
+
+<TopBlock>
+
+Today you'll be building an and deploying a Cloudflare Worker that uses Prisma to save every request to a MongoDB database for auditing later.
```
Comment (ajcwebdev) [inline content/300-guides/200-deployment/110-deployment-guides/450-deploying-to-cloudflare-workers.mdx:9]: Should be:

> Today you'll be building ~~an~~ and deploying

File: content/300-guides/200-deployment/110-deployment-guides/450-deploying-to-cloudflare-workers.mdx [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+- Free [MongoDB Atlas](https://www.mongodb.com/cloud/atlas) database with an accessible URL
+  - e.g. `mongodb+srv://<user>:<password>@cluster.mongodb.net`.
+- Free [Cloudflare Worker](https://workers.cloudflare.com/) account.
+- Free [GitHub](https://github.com/) account.
+- Node.js & NPM installed.
+- Git installed.
+
+## 1. Setup your Application
+
+Open your terminal and navigate to a location of your choice. Run the following to commands to setup your application.
```
Comment (ajcwebdev) [inline content/300-guides/200-deployment/110-deployment-guides/450-deploying-to-cloudflare-workers.mdx:26]: Should be:

> Run the following ~~to~~ commands to setup your application.

File: content/300-guides/200-deployment/110-deployment-guides/450-deploying-to-cloudflare-workers.mdx [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+request method: GET!
+```
+
+And if you check you're database, you should see info logs written to your database
+
+![MongoDB Compass](./450-07-mongodb-compass.png)
+
+It's working locally!
+
+## 11. Publishing to Cloudflare Workers
```
Comment (ajcwebdev) [inline content/300-guides/200-deployment/110-deployment-guides/450-deploying-to-cloudflare-workers.mdx:298]: ~~Should be: `## 12. Publishing to Cloudflare Workers`~~

EDIT: Correction, the list skips number `2`, so this should still be section 11 and all the others should be one less.

**Comment by ajcwebdev** [comment]
## Missing Webpack Steps?

I made it up to the `11. Develop the Cloudflare Worker function` section and hit a few blockers I could resolve and then eventually one that I couldn't.

### 1. `webpack_config` in `wrangler.toml`

First I got:

```
⚠️  If you would like to use your own custom webpack configuration, you will need to add this to your configuration file:
webpack_config = "webpack.config.js"
Error: The entrypoint of your Worker (index.js) could not be found.
```

### 2. Install `webpack` dependency

I added `webpack_config = "webpack.config.js"` to my `wrangler.toml` file. Then another error was thrown:

```
👀  ./src/index.ts
Module build failed (from ./node_modules/ts-loader/index.js):
Error: Cannot find module 'webpack'
Require stack:
- /Users/ajcwebdev/ajcwebdev-prisma-cloudflare/node_modules/ts-loader/dist/instances.js
- /Users/ajcwebdev/ajcwebdev-prisma-cloudflare/node_modules/ts-loader/dist/index.js
- /Users/ajcwebdev/ajcwebdev-prisma-cloudflare/node_modules/ts-loader/index.js
- /Users/ajcwebdev/Library/Caches/.wrangler/wranglerjs-1.19.3/node_modules/loader-runner/lib/loadLoader.js
- /Users/ajcwebdev/Library/Caches/.wrangler/wranglerjs-1.19.3/node_modules/loader-runner/lib/LoaderRunner.js
- /Users/ajcwebdev/Library/Caches/.wrangler/wranglerjs-1.19.3/node_modules/webpack/lib/NormalModule.js
- /Users/ajcwebdev/Library/Caches/.wrangler/wranglerjs-1.19.3/node_modules/webpack/lib/NormalModuleFactory.js
- /Users/ajcwebdev/Library/Caches/.wrangler/wranglerjs-1.19.3/node_modules/webpack/lib/Compiler.js
- /Users/ajcwebdev/Library/Caches/.wrangler/wranglerjs-1.19.3/node_modules/webpack/lib/webpack.js
- /Users/ajcwebdev/Library/Caches/.wrangler/wranglerjs-1.19.3/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.Module._load (internal/modules/cjs/loader.js:746:27)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:93:18)
    at Object.<anonymous> (/Users/ajcwebdev/ajcwebdev-prisma-cloudflare/node_modules/ts-loader/dist/instances.js:7:17)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Module.require (internal/modules/cjs/loader.js:974:19)
Error: webpack returned an error. Try configuring `entry` in your webpack config relative to the current working directory, or setting `context = __dirname` in your webpack config.
```

I added `webpack` to my dev dependencies.

```bash
yarn add -D webpack
npm i --save-dev webpack
```

## PrismaClient is unable to be run in the browser

After that it started throwing `PrismaClient is unable to be run in the browser`:

```
Error: Something went wrong with the request to Cloudflare...
Uncaught Error: PrismaClient is unable to be run in the browser.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
  at line 714 in PrismaClient
  at line 3150 in ./src/index.ts
  at line 20 in __webpack_require__
  at line 84
  at line 87
 [API code: 10021]
```

Interestingly enough, I tried moving `const prisma = new PrismaClient()` into the `handleRequest` function:

```ts
async function handleRequest(request: Request): Promise<Response> {
  const prisma = new PrismaClient()
  await prisma.log.create({
    data: {
      level: 'Info',
      message: `${request.method} ${request.url}`,
      meta: {
        headers: JSON.stringify(request.headers),
      },
    },
  })
  return new Response(`request method: ${request.method}!`)
}
```

And then it appeared to actually deploy:

```
👂  Listening on http://127.0.0.1:8787
🌀  Detected changes...
💁  Ignoring stale first change
```

Visiting the link resulted in the same error though in the terminal:

```
[2021-10-18 23:28:11] GET prisma-mongodb-cloudflare.anthonycampolo.workers.dev/ HTTP/1.1 500 Internal Server Error
Error: PrismaClient is unable to be run in the browser.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (worker.js:714:11)
    at handleRequest (worker.js:3154:20)
    at worker.js:3151:23 at line 713, col 9
{
  "exceptionDetails": {
    "columnNumber": 9,
    "exception": {
      "className": "Error",
      "description": "Error: PrismaClient is unable to be run in the browser.\nIn case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues\n    at new PrismaClient (worker.js:714:11)\n    at handleRequest (worker.js:3154:20)\n    at worker.js:3151:23",
      "preview": null,
      "subtype": "error",
      "type": "object",
      "value": null
    },
    "lineNumber": 713,
    "text": "Uncaught (in promise)",
    "url": "worker.js"
  },
  "timestamp": 1634624891969
}
Error: PrismaClient is unable to be run in the browser.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues at line 0, col -2
...
```

And the browser displayed:

Screen Shot 2021-10-18 at 11 28 18 PM

`npx wrangler publish` similarly was able to deploy:

```bash
✨  Built successfully, built project size is 22 KiB.
✨  Successfully published your script to
 https://prisma-mongodb-cloudflare.anthonycampolo.workers.dev
```

But [resulted in an error](https://prisma-mongodb-cloudflare.anthonycampolo.workers.dev).

Anyway, I'm assuming this is a massively complicated undertaking, and this was just pushed a few hours ago, so bugs were expected. Thanks for taking this on, I've been [hoping for Workers compatibility with Prisma for a long time](https://community.redwoodjs.com/t/running-redwoodjs-on-cloudflare-workers/2013) and I'm very excited to experiment with this.

## [Reproduction Repo](https://github.com/ajcwebdev/ajcwebdev-prisma-cloudflare)

EDIT: And one other thing I noticed was at the `Importing your Project into the Prisma Data Platform` section, I was getting this error:

```
Unable to link database: Error in connector: Error querying the database: Server selection timeout: No available servers.

Topology: {
  Type: ReplicaSetNoPrimary,
  Servers: [
    {
      Address: prisma-cloudflare-worke-shard-00-02.sl3jz.mongodb.net:27017,
      Type: Unknown, Error: unexpected end of file
    },
    {
      Address: prisma-cloudflare-worke-shard-00-00.sl3jz.mongodb.net:27017,
      Type: Unknown, Error: unexpected end of file
    },
    {
      Address: prisma-cloudflare-worke-shard-00-01.sl3jz.mongodb.net:27017,
      Type: Unknown, Error: unexpected end of file
    },
  ]
}
```

I created another project and the only thing I changed was instead of setting the IP address to my own when creating the connection string, I selected "Allow access from anywhere (0.0.0.0/0)" in the Network Permission List. That seemed to work, although I'm not sure if the change was actually the fix or just a coincidence.

File: content/300-guides/200-deployment/110-deployment-guides/450-deploying-to-cloudflare-workers.mdx [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+}
+```
+
+And install `@cloudflare/workers-types` with npm:
+
+```terminal
+npm install -D @cloudflare/workers-types ts-loader@8.3.0
+```
+
+<Admonition type="info">
```
Comment (nikolasburk) [inline content/300-guides/200-deployment/110-deployment-guides/450-deploying-to-cloudflare-workers.mdx:86]: The rendering of the Markdown link here seems to be broken: 
image

**Comment by ajcwebdev** [comment]
After upgrading to v3.3.0 I got a different error from Cloudflare:

```
Error: Schema (re)uploaded
    at Object.requestInternal (worker.js:918:435)
    at async cb (worker.js:919:6155)
    at async handleRequest (worker.js:960:5) at line 0, col -2
[2021-10-19 11:58:36] GET prisma-mongodb-cloudflare.anthonycampolo.workers.dev/ HTTP/1.1 500 Internal Server Error
{
  "exceptionDetails": {
    "columnNumber": -2,
    "exception": {
      "className": "Error",
      "description": "Error: Schema (re)uploaded\n    at Object.requestInternal (worker.js:918:435)\n    at async cb (worker.js:919:6155)\n    at async handleRequest (worker.js:960:5)",
      "preview": {
        "description": "Error: Schema (re)uploaded\n    at Object.requestInternal (worker.js:918:435)\n    at async cb (worker.js:919:6155)\n    at async handleRequest (worker.js:960:5)",
        "entries": null,
        "overflow": false,
        "properties": [
          {
            "name": "clientVersion",
            "subtype": null,
            "type": "string",
            "value": "3.3.0",
            "valuePreview": null
          },
          {
            "name": "stack",
            "subtype": null,
            "type": "string",
            "value": "Error: Schema (re)uploaded\n    at Object.requestIn…155)\n    at async handleRequest (worker.js:960:5)",
            "valuePreview": null
          },
          {
            "name": "message",
            "subtype": null,
            "type": "string",
            "value": "Schema (re)uploaded",
            "valuePreview": null
          }
        ],
        "subtype": "error",
        "type": "object"
      },
      "subtype": "error",
      "type": "object",
      "value": null
    },
    "lineNumber": 0,
    "text": "Uncaught (in promise)",
    "url": "undefined"
  },
  "timestamp": 1634669938764
}
Error: Schema (re)uploaded at line 0, col -2
```

### [Repo with current version](https://github.com/ajcwebdev/ajcwebdev-prisma-cloudflare)

**Pull request merged**